Cookbook 24.04 migration

Crola1702 opened the pull request:
# Description

This PR includes:
* Multiple updates to the Jenkins configuration recipe and plugins management for 24.04 migraion
* Improves the management of Jenkins authentication by removing any reference to old jenkins cookbook resources. Also, uses groovy scripts put in init.groovy.d directory for initializing authentication and configuring authorization
* Replaces the use of `jenkins_*_credentials` resources with Groovy scripts to handle the use of credentials more dynamically and flexibly.
* Updates the agents to work with Ubuntu 24.04 by:

## Changes:
* Updated Jenkins version to `2.479.2` 
* Added a new attribute for JDK version 
* Changed the included recipe from `jenkins::master` to `jenkins::jenkins`
* Updated numerous Jenkins plugins to their latest versions to ensure compatibility and security
* Improved the removal of obsolete plugins by using a filtered grep command to delete specific `.jpi` files.
* Added a new recipe to manage plugin installation using a `plugins.txt` template and the Jenkins plugin manager tool.
* Updated Jenkins configuration in `jenkins.yaml.erb` to reflect changes in plugin names and configurations.
* Created `/var/lib/jenkins/init.groovy.d` to store `auth_strategy.groovy` defined by the selected authorization strategy defined in the cookbook attributes
* Replaced the use of `jenkins_script` and `jenkins_user` to use groovy scripts for the auth_strategy
* Also, it creates `credentials_config.groovy` inside `/var/lib/jenkins/init.groovy.d/` (see #142)
* Configuring jdk version to be 21 for agents.
* Add bzip2 package in agents (not present by default in 24.04)

Crola1702 committed:
24.04 migration: Update Jenkins to latest LTS with updated plugins (#141)

* Update plugin install and removal

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Fix plugin_remove_filter undefined variable

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Upgrade publish-over-ssh to 1.24 in plugin list

This plugin had a security vulnerability, this is explained in https://github.com/ros-infrastructure/cookbook-ros-buildfarm/commit/263f7d281de79745574dc79b7f100b07d180ccb9.

Now https://plugins.jenkins.io/publish-over-ssh/ shows that security vulnerabilities were affecting versions 1.23 and earlier, so this is a bump to 1.24, where these vulnerabilities are not present anymore

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Update jenkins version

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Move plugins template file

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Upgrade openjdk version 8->21

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Update plugin list

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Apply suggested changes from  CasC plugin

* 'gitSCM' is an obsolete attribute name, please use 'scmGit'
* 'timestamperConfig' is an obsolete attribute name, please use 'timestamper'

* Parametrize jenkins jdk version from attributes

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Move plugin removal to plugins recipe

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

---------

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>
Co-authored-by: root <root@ip-10-0-1-10.us-west-2.compute.internal>

Crola1702 committed:
24.04 migration: Authentication Strategy (#142)

* Update plugin install and removal

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Move plugins template file

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Move plugin removal to plugins recipe

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Put authentication scripts and old resources in auth_strategy.groovy

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Fix syntax error

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Fix user creation script

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Use correct strategy in attributes and actually save the user

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Use mode 0500 for groovy scripts in init.groovy.d

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

---------

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
24.04 migration: Credentials (#143)

* Update plugin install and removal

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Fix plugin_remove_filter undefined variable

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Upgrade publish-over-ssh to 1.24 in plugin list

This plugin had a security vulnerability, this is explained in https://github.com/ros-infrastructure/cookbook-ros-buildfarm/commit/263f7d281de79745574dc79b7f100b07d180ccb9.

Now https://plugins.jenkins.io/publish-over-ssh/ shows that security vulnerabilities were affecting versions 1.23 and earlier, so this is a bump to 1.24, where these vulnerabilities are not present anymore

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Move plugins template file

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Upgrade openjdk version 8->21

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Move plugin removal to plugins recipe

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Fix syntax error

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Fix user creation script

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Fixed double variable declared

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Migrate credentials resources to groovy scripts

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Fix groovy syntax errors

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Remove credentials todos

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Use mode 0500 for groovy scripts in init.groovy.d

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

---------

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
24.04 migration: Agents (#145)

* Update plugin install and removal

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Fix plugin_remove_filter undefined variable

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Upgrade publish-over-ssh to 1.24 in plugin list

This plugin had a security vulnerability, this is explained in https://github.com/ros-infrastructure/cookbook-ros-buildfarm/commit/263f7d281de79745574dc79b7f100b07d180ccb9.

Now https://plugins.jenkins.io/publish-over-ssh/ shows that security vulnerabilities were affecting versions 1.23 and earlier, so this is a bump to 1.24, where these vulnerabilities are not present anymore

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Update jenkins version

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Move plugins template file

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Upgrade openjdk version 8->21

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Update plugin list

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Apply suggested changes from  CasC plugin

* 'gitSCM' is an obsolete attribute name, please use 'scmGit'
* 'timestamperConfig' is an obsolete attribute name, please use 'timestamper'

* Parametrize jenkins jdk version from attributes

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Move plugin removal to plugins recipe

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Put authentication scripts and old resources in auth_strategy.groovy

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Fix syntax error

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Fix user creation script

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Use correct strategy in attributes and actually save the user

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Fixed double variable declared

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Migrate credentials resources to groovy scripts

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Fix groovy syntax errors

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Remove credentials todos

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Agent: Parametrize java version

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Add bzip2 package dependecy

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Add bzip2 comment

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

---------

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>
Co-authored-by: root <root@ip-10-0-1-10.us-west-2.compute.internal>

Crola1702 committed:
Update to latest jenkins version

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Remove deprecated agentProtocols section

From: https://www.jenkins.io/doc/upgrade-guide/2.492/:

'The agentProtocols section of a configuration as code global configuration should be deleted because it will now be ignored. If you do not wish to allow inbound TCP agents, disable the port instead of the protocol.'

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Use environment override attributes

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Add jenkins to jenkins-agent group

Groovy system scripts use the master node user, i.e., jenkins (See https://github.com/ros-infrastructure/ros_buildfarm/blob/4a6c69587a04834fe99d6adbc3db124f2c6d109c/ros_buildfarm/templates/snippet/builder_system-groovy.xml.em#L1-L15).
This is a problem, as the reconfigure jobs need access to views under jenkins-agent user workspace

Everything under /home/jenkins-agent has 750 permissions, so jenkins user does not have access by default.
Adding jenkins to the group will give the jenkins admin access to the jenkins agent data.

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Upgrade jenkins to 492.2

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Approve signatures using groovy script (#149)

* Approve signatures using groovy script

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Use 99 to run approved_signatures at the end

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Fix acme.sh not_if clause

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* DEBUG: Comment approved-signatures script

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Restore approved_scripts.groovy

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Set order of init.groovy.d scripts

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Remove remotingSecurity

See: https://www.jenkins.io/doc/book/security/controller-isolation/jep-235/#api-compatibility

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Comment approved signatures script

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Stop clearingAPprovedSignatures before acutally approving some

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Remove scriptApproval attribute

We're using an approval_signatures.groovy script instead

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

---------

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

nuclearsandwich commented on attributes/jenkins.rb:
```diff
@@ -23,4 +23,7 @@
 #default['jenkins']['master']['version'] = '2.346.1'
 # Last version supporting sysvinit scripts
-default['jenkins']['master']['version'] = '2.319.3'
+default['jenkins']['master']['version'] = '2.492.2'
```
We'll eventually want to drop this pin entirely to allow unattended-upgrades to update Jenkins.

Crola1702 replied:
I opened an issue to track this into a different PR: See #158

nuclearsandwich commented on recipes/agent.rb:
```diff
@@ -111,6 +112,9 @@
 
 package 'python3-empy'
 
+# Install bzip2 as it's not present in ubuntu noble by default
```
Can you note what this dependency serves?

Crola1702 replied:
Done. It is used by https://github.com/ros-infrastructure/ros_buildfarm/blob/896e5e6cc90a0eb4fbfc2e31c22661e6adb08230/ros_buildfarm/templates/ci/ci_job.xml.em#L167 (added in a5158da)

nuclearsandwich commented on recipes/jenkins.rb:
```diff
@@ -455,13 +455,138 @@ def jenkins = Jenkins.getInstance()
 end
 
+file '/var/lib/jenkins/init.groovy.d/99-approved_signatures.groovy' do
+  content <<~GROOVY
```
This is a long enough set of contents that I'd consider making it a cookbook file (or even a template) rather than embedding the file contents here.

Crola1702 replied:
Done in 950603d

Crola1702 committed:
Fix typo in gpg private-key

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Fix typo in gpg private-key (again)

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

nuclearsandwich committed:
Add jenkins-agent to needrestart overrides.

needrestart in 24.04 now runs for both attended and unattended upgrades.
It has a built-in list of long-running services which should not be
restarted arbitrarily but need to include our jenkins-agent since it may
be running jobs and shouldn't be stopped by an unattended-upgrade run.

nuclearsandwich committed:
Prevent needrestart from restarting jenkins.

Jenkins is heavy enough that an inconvenient or unplanned restart is
quite disruptive.

nuclearsandwich committed:
Add a Groovy script for configuring a sensible default view.

Crola1702 committed:
Revert "Fix typo in gpg private-key (again)"

See https://github.com/osrf/chef-osrf/pull/303

This reverts commit 3773a05cf89c1a426b3ca581b5a10b1e989d9f4d.

Crola1702 committed:
Add noble in the apt_repos suite

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Fix acme.sh not_if clause

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Add Hudson model import

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Move approved-signatures script to cookbook_file

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Add missing production hashes

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Explicitly state why is bzip2 needed

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>
